File: octave/ecuaciones_no_lineales/punto_fijo.m
function [xk, k, error] = punto_fijo(f, a, b, x0, tol, iterMax)
    f1 = str2func(f); % 1    
    % f1 = matlabFunction(sym(f)); % 2

    k = 0;
    error = tol + 1;
    e = [];

    while (tol < error && k < iterMax)
        xk= f1(x0);
        error = abs(f1(xk) - xk);
        e = [e error];
        x0 = xk;
        ++k; 
    end
    
    plot(1 : k, e)
    xlabel('iter (k)')
    ylabel('Error (|f(x_k)|)')
    title('Error del Metodo del Punto Fijo')
end

% 1
% [xk, k, error] = punto_fijo('@(x)log(2*x+1)', 0.5, 1, 2, 10^-8, 5)

% 2
% [xk, k, error] = punto_fijo(log(2*x+1), 0.5, 1, 2, 10^-8, 5)

File: python/ecuaciones_no_lineales/punto_fijo.py
import math
import matplotlib.pyplot
from sympy import *


def punto_fijo(p, xo, a, b, tol, iterMax):
    x = Symbol('x')
    f1 = sympify(p)
    error = tol + 1
    k = 0
    it = []
    e = []

    while tol < error and k < iterMax:
        xk = f1.evalf(subs={x: xo})
        error = abs((f1.evalf(subs={x: xk})) - xk)
        e.append(error)
        it.append(k)
        xo = xk
        k += 1

    matplotlib.pyplot.plot(it, e)
    ans = {
        "x": xk,
        "k": k,
        "error": e[-1]
    }
    return ans

# punto_fijo('log(2*x+1)',0.5,1,2,10^-8,5)
